pkg/mutation/mutators/core/setter.go

```diff
@@ -0,0 +1,166 @@
+	if !ok {
+		return nil, fmt.Errorf("assign.value for keyed list is not an object: %+v", s.value)
+	}
+	return valueAsObject, nil
+}
+
+func (s *defaultSetter) SetValue(obj map[string]interface{}, key string) error {
+	value := s.value
+	incomingPlaceholder, isIncomingPlaceholder := types.ToExternalDataPlaceholder(value)
+	if _, ok := obj[key]; ok {
+		switch prev := obj[key].(type) {
+		case map[string]interface{}:
+			if previousPlaceholder, isPreviousPlaceholder := types.ToExternalDataPlaceholder(prev); isPreviousPlaceholder {
+				// replace old placeholder with new one
+				if isIncomingPlaceholder {
+					incomingPlaceholder.ValueAtLocation = previousPlaceholder.ValueAtLocation
+				} else if valueAsString, err := encodeToString(value); err == nil {
+					// assign incoming value as string to the previous placeholder
+					incomingPlaceholder = previousPlaceholder
+					incomingPlaceholder.ValueAtLocation = valueAsString
+				} else {
+					// return an error if the incoming value can't be encoded to a string
+					return err
+				}
+			}
+		default:
+			if isIncomingPlaceholder {
+				valueAsString, err := encodeToString(prev)
+				if err != nil {
+					return err
+				}
+				incomingPlaceholder.ValueAtLocation = valueAsString
+			}
+		}
+	}
+
+	var err error
+	if incomingPlaceholder != nil {
+		value, err = incomingPlaceholder.ToUnstructured()
+		if err != nil {
+			return fmt.Errorf("unable to convert external data placeholder to unstructured: %w", err)
+		}
+	}
+
+	obj[key] = value
+	return nil
+}
+
+func (s *defaultSetter) MergeListElements(old interface{}, new map[string]interface{}) (map[string]interface{}, error) {
+	if old == nil {
+		return new, nil
+	}
+
+	oldAsMap, ok := old.(map[string]interface{})
+	if !ok {
+		return new, nil
+	}
+
+	for k, oldValue := range oldAsMap {
+		if _, ok := new[k]; !ok {
+			new[k] = oldValue
+			continue
+		}
+
+		oldPlaceholder, isOldPlaceholder := types.ToExternalDataPlaceholder(oldValue)
+		newPlaceholder, isNewPlaceholder := types.ToExternalDataPlaceholder(new[k])
+		mergedPlaceholder := newPlaceholder
+
+		var value interface{}
+		switch {
+		case isOldPlaceholder && isNewPlaceholder:
+			value = oldPlaceholder.ValueAtLocation
+		case isOldPlaceholder:
+			value = new[k]
+			mergedPlaceholder = oldPlaceholder
+		case isNewPlaceholder:
+			value = oldValue
+		}
+
+		if mergedPlaceholder != nil {
+			valueAsString, err := encodeToString(value)
+			if err != nil {
+				return nil, err
+			}
+			mergedPlaceholder.ValueAtLocation = valueAsString
+
+			unstructured, err := mergedPlaceholder.ToUnstructured()
+			if err != nil {
+				return nil, err
+			}
+			new[k] = unstructured
+		}
+	}
+
+	return new, nil
+}
+
+func encodeToString(value interface{}) (string, error) {
+	switch value := value.(type) {
+	case string:
+		return value, nil
+	case int64:
+		return strconv.FormatInt(value, 10), nil
+	case float64:
+		return strconv.FormatFloat(value, 'f', -1, 64), nil
```

Comment:
Are we concerned about rounding errors with floats?

Reply:
Setting `prec` as `-1` will use the smallest number of digits necessary such that ParseFloat will return the original float exactly. according to https://pkg.go.dev/strconv#FormatFloat. Note that it's the external data provider's responsibility to parse the string-represented float back to float, process it, and send the resulting float back to us.

Reply:
Makes sense. What about the inbound path. Does K8s handle the conversion of floats losslessly? Does K8s use floats? IIRC mem/CPU limits can use floats embedded in a string, but are there any raw floats floating around?

Reply:
> Does K8s handle the conversion of floats losslessly?

We received the float from the external data provider as `interface{}` in the external data response and we directly assign it to the object without altering the result so I believe it's lossless.  The only gotcha I am aware of is that `json.Marshal` will remove the float's trailing zeros.

> Does K8s use floats / are there any raw floats floating around?

Great question. I actually don't see any float64 in any core API after doing a quick search in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/, but user-defined cards can have float.

For mem/CPU limits, as you mentioned, can be represented in strings so I wouldn't be too concerned about precision as long as the external data provider is sending the result in strings.

Reply:
> We received the float from the external data provider

Right, we have full control over our request path, more curious what the K8s API server would do if it encountered a float variable.

The main issue here is rounding errors coming from the fact that floating point numbers can't exactly represent decimal numbers, which could complicate cache comparison: https://www.tomasbeuzen.com/post/floating-point-numbers/#4

Here's what the K8s API guidelines have to say about this: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types (sounds like we wont see floats in the K8s API).

Also, we are deserializing to unstructured, not custom types. Might be worth experimenting to see how the API server handles floats to see if it's a concern before we commit to dealing with the vagaries.

Reply:
Makes sense. I can remove float from this function i.e. we won't support float as outbound data to external data providers until we experiment with it.